File: quorum/agy_watch.py
"""Daemon thread that tails agy.log and fires teardown on a rate-limit signal.

agy.log is the only deterministic continuous rate-limit signal: gauntlet does
not stream the agy tmux pane, so polling the log is the sole way to detect
RESOURCE_EXHAUSTED / 429 while the agy run is in flight.
"""
from __future__ import annotations

import threading
import time
from pathlib import Path

from quorum.agy_teardown import kill_run_tmux_server
from quorum.runner import _agy_log_shows_rate_limit


class AgyRateLimitWatcher(threading.Thread):
    """Poll *log_path* for rate-limit signals and call *teardown* on first hit.

    Tolerates the log file being absent at start — agy creates it late.
    """

    def __init__(
        self,
        log_path: Path,
        scratch_dir: Path,
        *,
        teardown=None,
        poll_interval: float = 0.5,
    ) -> None:
        super().__init__(daemon=True)
        self._log_path = Path(log_path)
        self._scratch_dir = scratch_dir
        self._teardown = teardown if teardown is not None else kill_run_tmux_server
        self._poll_interval = poll_interval
        self._stop_event = threading.Event()

        # Public state — written before tripped is set to True so readers that
        # check tripped always find matched_text already populated.
        self.matched_text: str = ""
        self.tripped: bool = False

    def run(self) -> None:
        offset = 0
        while not self._stop_event.is_set():
            try:
                if self._log_path.exists():
                    with self._log_path.open("rb") as fh:
                        fh.seek(offset)
                        new_bytes = fh.read()
                    if new_bytes:
                        new_text = new_bytes.decode("utf-8", errors="replace")
                        offset += len(new_bytes)
                        if _agy_log_shows_rate_limit(new_text):
                            self.matched_text = new_text
                            self._teardown(self._scratch_dir)
                            # Set flag last so readers always see matched_text
                            # and teardown side-effect before tripped=True.
                            self.tripped = True
                            return
            except OSError:
                pass  # file disappeared mid-read; keep waiting
            self._stop_event.wait(self._poll_interval)

    def stop(self, timeout: float = 2.0) -> None:
        """Signal the polling loop to exit and join the thread."""
        self._stop_event.set()
        self.join(timeout=timeout)

File: src/agents/agy-watch.ts
import { closeSync, existsSync, fstatSync, openSync, readSync } from 'node:fs';
import { killRunTmuxServer } from './agy-teardown.ts';

// Async watcher that tails agy.log and fires teardown on a rate-limit signal.
//
// agy.log is the only deterministic continuous rate-limit signal: gauntlet does
// not stream the agy tmux pane, so polling the log is the sole way to detect
// RESOURCE_EXHAUSTED / 429 while the agy run is in flight.
//
// Port of quorum/agy_watch.py. The Python version is a threading.Thread daemon;
// the TS version is a class with start()/stop() backed by an async poll loop
// (not a busy thread). Semantics preserved: fire teardown once, clean stop, no
// leaked timer.

// Substrings agy writes to its log/stderr when the Gemini Code Assist backend
// throttles. RESOURCE_EXHAUSTED is the definitive 429 signal; ratelimitexceeded
// corroborates. Matched case-insensitively. Mirrors _AGY_RATE_LIMIT_SUBSTRINGS.
const AGY_RATE_LIMIT_SUBSTRINGS = ['resource_exhausted', 'ratelimitexceeded'];
// A bare "429" matches hex trace IDs (e.g. 0x...e4291...), ports, and byte
// counts that pepper agy's streaming log, false-tripping the mid-run watcher
// into killing a healthy run. Require a word-boundaried HTTP-status 429 instead.
const AGY_429_RE = /\b429\b/;

/**
 * True if any of *texts* contains an unambiguous Code Assist rate-limit signal.
 *
 * Pure predicate. Joins truthy texts with newlines, lowercases, and matches the
 * rate-limit substrings or a word-boundaried 429. Port of
 * quorum/runner._agy_log_shows_rate_limit.
 */
export function agyLogShowsRateLimit(...texts: string[]): boolean {
  const blob = texts
    .filter((t) => t !== '')
    .join('\n')
    .toLowerCase();
  if (AGY_RATE_LIMIT_SUBSTRINGS.some((sig) => blob.includes(sig))) {
    return true;
  }
  return AGY_429_RE.test(blob);
}

export type TeardownFn = (target: string) => unknown;

export interface AgyRateLimitWatcherOptions {
  /** Injectable teardown callback. Defaults to killRunTmuxServer. */
  readonly teardown?: TeardownFn;
  /** Poll interval in milliseconds. Defaults to 500 (Python 0.5s). */
  readonly pollIntervalMs?: number;
}

/**
 * Poll *logPath* for rate-limit signals and call *teardown* on the first hit.
 *
 * Tolerates the log file being absent at start — agy creates it late. Assumes a
 * single append-only run log; log rotation / truncation under the watcher is out
 * of scope and would stall detection — acceptable because each run gets its own
 * fresh log.
 */
export class AgyRateLimitWatcher {
  private readonly logPath: string;
  // Opaque arg handed to teardown(); the default teardown expects the run dir
  // and globs to the scratch dir itself, so this is not a scratch dir.
  private readonly teardownTarget: string;
  private readonly teardown: TeardownFn;
  private readonly pollIntervalMs: number;

  private stopRequested = false;
  private loopPromise: Promise<void> | null = null;
  private wakeUp: (() => void) | null = null;

  /** Public state — written before `tripped` is set so readers always see it. */
  matchedText = '';
  tripped = false;

  constructor(
    logPath: string,
    teardownTarget: string,
    opts: AgyRateLimitWatcherOptions = {},
  ) {
    this.logPath = logPath;
    this.teardownTarget = teardownTarget;
    this.teardown =
      opts.teardown ?? ((target: string) => killRunTmuxServer(target));
    this.pollIntervalMs = opts.pollIntervalMs ?? 500;
  }

  /** Whether the poll loop is currently active. */
  get running(): boolean {
    return this.loopPromise !== null;
  }

  /** Start the poll loop. Idempotent — a second call while running is a no-op. */
  start(): void {
    if (this.loopPromise !== null) {
      return;
    }
    this.stopRequested = false;
    this.loopPromise = this.runLoop();
  }

  /**
   * Signal the poll loop to exit and await it. Safe to call even if the watcher
   * was never started, and safe to call twice.
   */
  async stop(): Promise<void> {
    this.stopRequested = true;
    if (this.wakeUp) {
      this.wakeUp(); // wake an in-progress interval wait so stop returns promptly
    }
    const p = this.loopPromise;
    if (p) {
      await p;
    }
  }

  private async runLoop(): Promise<void> {
    let offset = 0;
    try {
      while (!this.stopRequested) {
        try {
          if (existsSync(this.logPath)) {
            const { text: newText, bytesRead } = this.readFrom(offset);
            offset += bytesRead;
            if (newText !== '' && agyLogShowsRateLimit(newText)) {
              this.matchedText = newText;
              this.teardown(this.teardownTarget);
              // Set flag last so readers always see matchedText and the
              // teardown side-effect before tripped=true.
              this.tripped = true;
              return;
            }
          }
        } catch {
          // file disappeared mid-read; keep waiting
        }
        await this.waitInterval();
      }
    } finally {
      this.loopPromise = null;
    }
  }

  // Read the bytes appended past *offset* and decode them (utf-8, lossy on
  // partial multibyte tails, matching Python's errors="replace"). Returns the
  // raw bytes consumed so the caller advances the offset by bytes, not chars.
  private readFrom(offset: number): { text: string; bytesRead: number } {
    const fd = openSync(this.logPath, 'r');
    try {
      const size = fstatSync(fd).size;
      const toRead = size - offset;
      if (toRead <= 0) {
        return { text: '', bytesRead: 0 };
      }
      const buf = Buffer.alloc(toRead);
      const read = readSync(fd, buf, 0, toRead, offset);
      return { text: buf.subarray(0, read).toString('utf8'), bytesRead: read };
    } finally {
      closeSync(fd);
    }
  }

  /** Sleep for the poll interval, returning early if stop() wakes us. */
  private waitInterval(): Promise<void> {
    return new Promise<void>((resolve) => {
      const timer = setTimeout(() => {
        this.wakeUp = null;
        resolve();
      }, this.pollIntervalMs);
      this.wakeUp = () => {
        clearTimeout(timer);
        this.wakeUp = null;
        resolve();
      };
    });
  }
}
